cmd/lakectl/cmd/retry_client.go

```diff
@@ -0,0 +1,72 @@
+package cmd
+
+import (
+	"context"
+	"crypto/x509"
+	"net/http"
+	"net/url"
+	"regexp"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+var (
+	// these errors aren't typed, so we match by regexp
+	redirectsErrorRe  = regexp.MustCompile(`stopped after \d+ redirects\z`)
+	schemeErrorRe     = regexp.MustCompile(`unsupported protocol scheme`)
+	notTrustedErrorRe = regexp.MustCompile(`certificate is not trusted`)
+)
+
+func NewRetryClient(maxRetries int, retryWaitMin, retryWaitMax time.Duration, transport *http.Transport) *http.Client {
+	retryClient := retryablehttp.NewClient()
+	if transport != nil {
+		retryClient.HTTPClient.Transport = transport
+	}
+	retryClient.RetryMax = maxRetries
+	retryClient.RetryWaitMin = retryWaitMin
+	retryClient.RetryWaitMax = retryWaitMax
+	retryClient.CheckRetry = lakectlRetryPolicy
+	return retryClient.StandardClient()
+}
+
+// lakectl retry policy - we retry in the following cases:
+// HTTP status 429 - too many requests
+// HTTP status 500 - internal server error - could be recoverable
+// HTTP status 503 - service unavailable
+// We retry on all client transport errors except for:
+//   - too many redirects
+//   - invalid http scheme/protocol
+//   - TLS cert verification failure
+func lakectlRetryPolicy(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	// do not retry on context.Canceled or context.DeadlineExceeded
+	if ctx.Err() != nil {
+		return false, ctx.Err()
+	}
+
+	// handle client transport errors
+	if err != nil {
+		if v, ok := err.(*url.Error); ok {
+			if redirectsErrorRe.MatchString(v.Error()) || // too many redirects
```

Comment:
I'm very confused by this code: how can package url every return "too many redirects"? It is literally only a URL parser, and I thought redirects happened after that?
(Also for the other 2 patterns, I guess?)

Reply:
I agree that it's abusing the `url.Error` type, but looks like that's what they went with... [ref](https://github.com/golang/go/blob/master/src/net/http/client.go#L69).  
It would be much nicer if these errors were specifically typed so we could use `errors.Is`.

Reply:
This is 100% true but NOT FUNNY.  I spent 3 minutes down that rabbit hole and ran back out screaming.

Reply:
We _could_ implement our own callback and return a typed error, but I feel like the ROI will be very low.